metashare/api/jobs.py

```diff
@@ -293,3 +297,47 @@ def refresh_github_repositories_for_user(user):
 
 
 refresh_github_repositories_for_user_job = job(refresh_github_repositories_for_user)
+
+
+# This avoids partially-applied saving:
+@transaction.atomic
+def refresh_commits(*, repository, branch_name):
+    """
+    This should only run when we're notified of a force-commit. It's the
+    nuclear option.
+    """
+    from .models import User, Project, Task
+
+    user = User.objects.get(pk=settings.GITHUB_USER_ID)
+    repo = get_repo_info(user, repository.repo_id)
+    commits = repo.commits(repo.branch(branch_name).latest_sha())
+
+    matching_self = repository.branch_name == branch_name
+    if matching_self:
+        repository.commits = [normalize_commit(c) for c in commits]
+        repository.finalize_repository_update()
+
+    projects = Project.objects.filter(repository=repository, branch_name=branch_name)
+    for project in projects:
+        parent_commits_set = project.repository.commit_set
+        project.commits = [
+            normalize_commit(commit)
+            for commit in commits
+            if commit.sha not in parent_commits_set
+        ]
+        project.finalize_project_update()
+
+    tasks = Task.objects.filter(project__repository=repository, branch_name=branch_name)
+    for task in tasks:
+        parent_commits_set = (
+            task.project.commit_set | task.project.repository.commit_set
+        )
+        task.commits = [
+            normalize_commit(commit)
+            for commit in commits
+            if commit.sha not in parent_commits_set
+        ]
```

Comment:
Instead of comparing against project commits, could this fetch all commits back to (and including) the commit that created the branch?